### PR TITLE
ADR 012: Multi-Version Support

### PR DESCRIPTION
**Pull request by mzabaluev**
[Rendered](https://github.com/informalsystems/tendermint-rs/blob/mikhail/adr-012-multi-version/docs/architecture/adr-012-multi-version.md)

### CODE REVIEW

File: docs/architecture/adr-012-multi-version.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+This crate is not actively supported and we should only make minimal effort
+to make it work with the proposed changes in a single chosen version of the
+protocol.
+
+## Open questions
+
+* Should we migrate tendermint-abci to 0.37, freeze it at 0.34 and declare
+  end of life, or commit to the same level of multi-version support as in
+  other crates?
```
Comment (romac) [inline docs/architecture/adr-012-multi-version.md:104]: As discussed during the sync review, we've decided to migrate it to 0.37 in order to give its eventual customers a chance to update their apps for 0.37. We will eventually deprecate `tendermint-abci` in favor of [`tower-abci`](https://github.com/penumbra-zone/tower-abci) once the latter adds support for 0.37.

Let's update the ADR with this answer?

Comment (mzabaluev) [reply docs/architecture/adr-012-multi-version.md:104]: OK. I'll add a negative consequence then: the existing `tendermint-abci` users who wish to remain on 0.34 will be effectively stranded on the previous release (0.29.x currently) of the crate.

Comment (romac) [reply docs/architecture/adr-012-multi-version.md:104]: Good point!

File: docs/architecture/adr-012-multi-version.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+in the long term, there is only an ad-hoc way to discover which version is used
+by the node and select the appropriate encoding. It appears to be too
+late in the development process to introduce significant changes to mitigate
+this in the 0.37 release timeframe. A requirement to specify the
+protocol version in chain configuration would complicate the deployment
+experience for chain and relayer operators and should be avoided.
+
+Another motivation for multi-version support is versioning of tendermint-rs
+itself: previously, the semver compatibility of tendermint-rs releases was
+tied with corresponding CometBFT Core version sequences in a none too obvious
```
Comment (thanethomson) [inline docs/architecture/adr-012-multi-version.md:35]:
```suggestion
tied with corresponding CometBFT version sequences in a none too obvious
```

File: docs/architecture/adr-012-multi-version.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+way. It's desirable to decouple tendermint-rs versioning from that of CometBFT
+Core, making it solely a matter of Rust crate API evolution. In the long term,
```
Comment (thanethomson) [inline docs/architecture/adr-012-multi-version.md:37]:
```suggestion
way. It's desirable to decouple tendermint-rs versioning from that of CometBFT,
making it solely a matter of Rust crate API evolution. In the long term,
```

File: docs/architecture/adr-012-multi-version.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+a single version of the tendermint-rs libraries should be able to support all
+versions of CometBFT Core protocols that are relevant to the community.
```
Comment (thanethomson) [inline docs/architecture/adr-012-multi-version.md:39]:
```suggestion
versions of CometBFT protocols that are relevant to the community.
```

File: docs/architecture/adr-012-multi-version.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+The developers of CometBFT protocols should take versioning practices into use
+for future revisions that support backward compatibility on the source code level.
+
+## Decision
+
+The tendermint-rs library API will be modified to provide support for
+multiple versions of CometBFT Core protocol, at this moment being 0.34 and
```
Comment (thanethomson) [inline docs/architecture/adr-012-multi-version.md:46]:
```suggestion
multiple versions of CometBFT APIs, at this moment being 0.34 and
```

File: docs/architecture/adr-012-multi-version.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+discover the version when needed, and have the client exhibit interior
+mutability that is otherwise not needed. In the future, a more formalized way
+to discover the protocol version in use should be provided, so this ad-hoc
+approach is expected to be eventually deprecated.
+
+### tendermint-abci
+
+This crate is not actively supported and we should only make the minimal effort
+to update it to the 0.37 version of ABCI. No backward compatibility with 0.34
+or multi-versoin support is to be implemented; the consumers should be steered
```
Comment (thanethomson) [inline docs/architecture/adr-012-multi-version.md:110]:
```suggestion
or multi-version support is to be implemented; the consumers should be steered
```

**Review by thanethomson — COMMENTED** [review]
Please include concrete examples of how the new API is to be used.

**Comment by mzabaluev** [comment]
Thank you @thanethomson, I have committed your suggestions.

> Please include concrete examples of how the new API is to be used.

Will do.

**Review by romac — APPROVED** [review]

**Pull request merged**